modin/engines/base/io/column_stores/parquet_dispatcher.py

```diff
@@ -79,58 +78,22 @@ def _read(cls, path, engine, columns, **kwargs):
-                meta = pd.metadata
-                column_names = pd.schema.to_arrow_schema().names
-            elif isinstance(path, str) and path.startswith("hdfs://"):
-                import fsspec.core
 
-                fs, path = fsspec.core.url_to_fs(path)
-                pd = ParquetDataset(path, filesystem=fs)
-                meta = pd.metadata
-                column_names = pd.schema.to_arrow_schema().names
-            elif isinstance(path, s3fs.S3File) or (
-                isinstance(path, str) and path.startswith("s3://")
-            ):
-                from botocore.exceptions import NoCredentialsError
+        if not columns:
+            import fsspec.core
+            from pandas.io.common import is_fsspec_url
 
-                if isinstance(path, s3fs.S3File):
-                    bucket_path = path.url().split(".s3.amazonaws.com")
-                    path = "s3://" + bucket_path[0].split("://")[1] + bucket_path[1]
-                try:
-                    fs = s3fs.S3FileSystem()
-                    pd = ParquetDataset(path, filesystem=fs)
-                except NoCredentialsError:
-                    fs = s3fs.S3FileSystem(anon=True)
-                    pd = ParquetDataset(path, filesystem=fs)
-                meta = pd.metadata
-                column_names = pd.schema.to_arrow_schema().names
-            else:
-                meta = ParquetFile(path).metadata
-                column_names = meta.schema.to_arrow_schema().names
+            fs, path_ = (
+                fsspec.core.url_to_fs(path, **(kwargs["storage_options"] or {}))
+                if is_fsspec_url(path)
+                else (None, path)
+            )
 
-            if meta is not None and meta.metadata is not None:
-                pandas_metadata = meta.metadata.get(b"pandas", None)
-                if pandas_metadata is not None:
-                    import json
+            dataset = ParquetDataset(path_, filesystem=fs, use_legacy_dataset=False)
+            column_names = dataset.schema.names
 
-                    # This is how we convert the metadata from pyarrow to a python
-                    # dictionary, from which we then get the index columns.
-                    # We use these to filter out from the columns in the metadata since
-                    # the pyarrow storage has no concept of row labels/index.
-                    # This ensures that our metadata lines up with the partitions without
-                    # extra communication steps once we have done all the remote
-                    # computation.
-                    index_columns = json.loads(pandas_metadata.decode("utf8")).get(
-                        "index_columns", []
-                    )
-                    column_names = [c for c in column_names if c not in index_columns]
+            if dataset.schema.pandas_metadata is not None:
+                index_columns = dataset.schema.pandas_metadata.get("index_columns", [])
```

Comment:
```suggestion
                index_columns = set(dataset.schema.pandas_metadata.get("index_columns", []))
```
to make next line a trifle faster :)

Reply:
`dataset.schema.pandas_metadata.get("index_columns", [])` is `dict` or `list`. In case `dict` we can't use that, because `dict` is unhashable.

Reply:
Uhm... :) unhashable means something if you want to use it as an _element_, but here it's being `converted` from a dictionary to a set. Works just fine:
```
>>> set({'a':1,'b':2})
{'b', 'a'}
```

Reply:
I introduced some misunderstanding. The result of `dataset.schema.pandas_metadata.get("index_columns", [])` is:
1. `list` of values. For example `['col0', 'col1']`. We can easy create `set` from this list.
2. `list` of dicts. For example `[{'kind': 'range', 'name': None, 'start': 0, 'stop': 256, 'step': 1}]`. In that case we cann't create set:
```python
l = [{'a': 1}]
s = set(l) # <-- TypeError: unhashable type: 'dict'
```